src/main/cpp/src/histogram.cu

```diff
@@ -0,0 +1,433 @@
+  auto const sizes_itr = thrust::make_constant_iterator(num_percentages);
+  auto offsets = std::get<0>(
+      cudf::detail::make_offsets_child_column(sizes_itr, sizes_itr + num_histograms, stream, mr));
+  auto output = cudf::make_lists_column(num_histograms, std::move(offsets), std::move(input),
+                                        null_count, std::move(null_mask), stream, mr);
+  if (null_count > 0) {
+    return cudf::detail::purge_nonempty_nulls(output->view(), stream, mr);
+  }
+
+  return output;
+}
+
+} // namespace
+
+std::unique_ptr<cudf::column> create_histogram_if_valid(cudf::column_view const &values,
+                                                        cudf::column_view const &frequencies,
+                                                        bool output_as_lists,
+                                                        rmm::cuda_stream_view stream,
+                                                        rmm::mr::device_memory_resource *mr) {
+  CUDF_EXPECTS(!frequencies.has_nulls(), "The input frequencies must not have nulls.",
+               std::invalid_argument);
+  CUDF_EXPECTS(frequencies.type().id() == cudf::type_id::INT64,
+               "The input frequencies must be of type INT64.", std::invalid_argument);
+  CUDF_EXPECTS(values.size() == frequencies.size(),
+               "The input values and frequencies must have the same size.", std::invalid_argument);
+
+  if (values.size() == 0) {
+    if (output_as_lists) {
+      return cudf::make_lists_column(
+          0, cudf::make_empty_column(cudf::type_to_id<cudf::size_type>()),
+          cudf::reduction::detail::make_empty_histogram_like(values), 0, {});
+    } else {
+      return cudf::reduction::detail::make_empty_histogram_like(values);
+    }
+  }
+
+  auto const default_mr = rmm::mr::get_current_device_resource();
+
+  // We only check if there is any row in frequencies that are negative (invalid) or zero.
+  auto check_invalid_and_zero =
+      cudf::detail::make_zeroed_device_uvector_async<int8_t>(2, stream, default_mr);
+
+  // We need to check and remember which rows are valid (positive) so we can do filtering later on.
+  auto check_valid = rmm::device_uvector<int8_t>(frequencies.size(), stream, default_mr);
+
+  thrust::for_each_n(
+      rmm::exec_policy(stream), thrust::make_counting_iterator(0), frequencies.size(),
+      [frequencies = frequencies.begin<int64_t>(), check_invalid = check_invalid_and_zero.begin(),
+       check_zero = check_invalid_and_zero.begin() + 1,
+       check_valid = check_valid.begin()] __device__(auto const idx) {
+        if (frequencies[idx] < 0) {
+          *check_invalid = 1;
+        }
+        if (frequencies[idx] == 0) {
+          *check_zero = 1;
+        }
+
+        check_valid[idx] = static_cast<int8_t>(frequencies[idx] > 0);
+      });
+
+  auto const h_checks = cudf::detail::make_std_vector_sync(check_invalid_and_zero, stream);
+  CUDF_EXPECTS(!h_checks.front(), // check invalid (negative) frequencies
```

Comment:
Because the check values are initialized with `0`, thus are guaranteed to be `0` if no negative frequencies exist:
```
auto check_invalid_and_zero =
      cudf::detail::make_zeroed_device_uvector_async<int8_t>(2, stream, default_mr);
```